### PR TITLE
LOG-2426: Vector: Support loki bearer token auth

### PR DESCRIPTION
**Pull request by vimalk78**
### Description
Cherry-pick of [LOG-2355](https://issues.redhat.com/browse/LOG-2355)

/cc @jcantrill 
/assign 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-2426

### CODE REVIEW

File: internal/generator/vector/output/loki/loki.go [hunk 1]
```diff
@@ -233,3 +235,40 @@ func BasicAuth(o logging.OutputSpec, secret *corev1.Secret) []Element {
...
+	conf := []Element{}
+	if o.Secret != nil {
+		// external loki, use user provided secret
+		// the secret contains the path of the token
+		if security.HasBearerTokenFileKey(secret) {
+			conf = append(conf, BasicAuthConf{
+				Desc:        "Bearer Auth Config",
+				ComponentID: strings.ToLower(vectorhelpers.Replacer.Replace(o.Name)),
+			})
+			token, _ := ioutil.ReadFile(security.GetFromSecret(secret, constants.BearerTokenFileKey))
```
Comment (syedriko) [inline internal/generator/vector/output/loki/loki.go:249]: it'd be better to handle the error here

File: internal/generator/vector/output/loki/bearer-token.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package loki
```
Comment (syedriko) [inline internal/generator/vector/output/loki/bearer-token.go:1]: underscores in file names is the golang way

File: internal/generator/vector/output/loki/loki-bearer-token-file [hunk 3]
```diff
@@ -0,0 +1 @@
+token-from-file
```
Comment (syedriko) [inline internal/generator/vector/output/loki/loki-bearer-token-file:1]: not a Go source, but may be stick to underscores, too?

File: internal/generator/vector/output/loki/loki_conf_test.go [hunk 4]
```diff
@@ -181,6 +182,96 @@ strategy = "basic"
...
 
+`,
+		}),
+		Entry("with bearer token", generator.ConfGenerateTest{
+			CLFSpec: logging.ClusterLogForwarderSpec{
+				Outputs: []logging.OutputSpec{
+					{
+						Type: logging.OutputTypeLoki,
+						Name: "loki-receiver",
+						URL:  "https://logs-us-west1.grafana.net",
```
Comment (syedriko) [inline internal/generator/vector/output/loki/loki_conf_test.go:193]: if this is a live endpoint, is a good idea to rely on it in our tests?
if it's not (dig is not resolving it for me), can we stick to example.com?

File: internal/generator/vector/outputs.go [hunk 5]
```diff
@@ -25,7 +27,18 @@ func Outputs(clspec *logging.ClusterLoggingSpec, secrets map[string]*corev1.Secr
 	ofp := OutputFromPipelines(clfspec, op)
 
 	for _, o := range clfspec.Outputs {
-		secret := secrets[o.Name]
+		var secret *corev1.Secret
+		if s, ok := secrets[o.Name]; ok {
+			secret = s
+			log.V(9).Info("Using secret configured in output")
```
Comment (syedriko) [inline internal/generator/vector/outputs.go:33]: when looking at the log, would it possible to tell which secret and which output we're talking about?

File: internal/generator/vector/outputs.go [hunk 6]
```diff
@@ -25,7 +27,18 @@ func Outputs(clspec *logging.ClusterLoggingSpec, secrets map[string]*corev1.Secr
...
+		} else {
+			secret = secrets[constants.LogCollectorToken]
+			if secret != nil {
+				log.V(9).Info("Using secret configured in logcollector-token")
```
Comment (syedriko) [inline internal/generator/vector/outputs.go:37]: "Using secret configured in " + constants.LogCollectorToken ?

File: internal/generator/vector/outputs.go [hunk 7]
```diff
@@ -25,7 +27,18 @@ func Outputs(clspec *logging.ClusterLoggingSpec, secrets map[string]*corev1.Secr
...
+			} else {
+				log.V(9).Info("No Secret found in logcollector-token")
```
Comment (syedriko) [inline internal/generator/vector/outputs.go:39]: "No secret found in " +  constants.LogCollectorToken ?

File: internal/generator/vector/output/loki/loki.go (hunk 1 above)
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki.go:249]: will add this to the verify step before conf generation

File: internal/generator/vector/output/loki/bearer-token.go (hunk 2 above)
Comment (vimalk78) [reply internal/generator/vector/output/loki/bearer-token.go:1]: sure

Comment (vimalk78) [reply internal/generator/vector/output/loki/bearer-token.go:1]: btw, there are many other files i have named with `-` , need to change them all to `_` ?

File: internal/generator/vector/output/loki/loki_conf_test.go (hunk 4 above)
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki_conf_test.go:193]: this is not a live endpoint, just for unit testing generating configuration

File: internal/generator/vector/outputs.go (hunk 5 above)
Comment (vimalk78) [reply internal/generator/vector/outputs.go:33]: updated

File: internal/generator/vector/outputs.go (hunk 7 above)
Comment (vimalk78) [reply internal/generator/vector/outputs.go:39]: updated

File: internal/generator/vector/output/loki/bearer-token.go (hunk 2 above)
Comment (vimalk78) [reply internal/generator/vector/output/loki/bearer-token.go:1]: updated

File: internal/generator/vector/output/loki/loki.go (hunk 1 above)
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki.go:249]: handled

File: internal/generator/vector/output/loki/loki_conf_test.go (hunk 4 above)
Comment (syedriko) [reply internal/generator/vector/output/loki/loki_conf_test.go:193]: can we change it to to example.com though so the intent is clear?

File: internal/generator/vector/output/loki/loki.go [hunk 8]
```diff
@@ -233,3 +235,47 @@ func BasicAuth(o logging.OutputSpec, secret *corev1.Secret) []Element {
...
+		if security.HasBearerTokenFileKey(secret) {
+			conf = append(conf, BasicAuthConf{
+				Desc:        "Bearer Auth Config",
+				ComponentID: strings.ToLower(vectorhelpers.Replacer.Replace(o.Name)),
+			})
+			var token string
+			token_file := security.GetFromSecret(secret, constants.BearerTokenFileKey)
+			token_bytes, err := ioutil.ReadFile(token_file)
+			if err != nil {
+				token = fmt.Sprintf("Error occurred. Reading file %s failed.", token_file)
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki.go:253]: there is nothing else that can be done

[the remark this answers is not in the document]
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki.go:253]: yes it is odd, and also incorrect, needs some discussion here.

The conf generator is a translator, it does not know how to handle errors. Thats is why we have a verify step before calling the generator. thats where all the checks are done prior to generating conf. e.g parsing URLs, and invariants etc.

This one is different. In this case, the secret contains a filepath, the file contains the bearer token. But that path is mounted in the collector pod, and conf generation is in CLO pod. So some change is needed here anyway.

[the remark this answers is not in the document]
Comment (vimalk78) [reply internal/generator/vector/output/loki/loki.go]: updated

**Review by jcantrill — COMMENTED** [review]
/lgtm

**Pull request merged**